### DISCUSSION TITLE
Parallel unsteady write error at 50 outputs

### DISCUSSION THREAD

**Discussion by LordOfBunnies** [CGNS/CGNS#327] [CGNStalk]
I'm currently working on my PhD at the University of Cincinnati.  I'm writing a lattice Boltzmann flow solver with the use of Amrex to do the adaptive mesh refinement.  I'm using CGNS for output (Amrex has its own output, but I can't get visualizers to work for it for some reason).

The problem I'm having is when doing many outputs, even with a static grid.  At 50 outputs, CGNS errors out and gives the following error:

(program name): H5ACmpio.c:1100: H5AC__log_moved_entry: Assertion '!entry_dirty' failed.

It does this regardless of how big the .cgns file is at the time, and regardless of what I'm running.

I've looked online, but I'm unable to find a good answer for why this is.  I'm guessing it's something simple, but not sure if it's hard coded (trying to grep for 50 leads to things like MAX_STACKS).  I'm not sure how to handle this within CGNS.

I'm currently using parallel CGNS 4.1.2 with parallel hdf5 1.12.0 on OpenMPI 1.10.7

**Comment by brtnfld** [comment]
That is an "ancient" version of OpenMPI, I don't think HDF5 (at least not 1.12.0) is tested with that version. Can you try a newer version? You could try OMPIO and see if that helps, mpirun --mca io ompio

**Reply by brtnfld** [reply]
@gsjaardema, you reported a similar issue with NetCDF4 and openmpi-4.02. and openmpi-1.10.2. Were you able to determine the issue?

**Reply by LordOfBunnies** [reply]
I'll try to get my department IT guy to upgrade to something relatively modern, but it might take a few days.  It's not horribly urgent since I can simply space out my outputs, but it would be nice to resolve it going forward.

**Comment by brtnfld** [comment]
> The problem I'm having is when doing many outputs, even with a static grid. At 50 outputs, CGNS errors out and gives the following error:

Do you mean after 50 open and closing of different CGNS files is when this error occurs? 

If switching to a newer version does not fix the problem, could you provide a reproducer?

**Reply by LordOfBunnies** [reply]
I do open and close the CGNS file every time I write data to it.  Is that a problem?  I mostly ask because I'm still writing things into my code so I like to see results if something causes it to bomb out.

**Reply by brtnfld** [reply]
Do you mean open/close for a check-point scenario (i.e., each time step), or do you mean after each write of several variables during a single checkpoint, for example? If it is the second case, then yes, that could be an issue and a performance issue because closing and opening a file carries some overhead usually associated with metadata initiation and clean-up

**Reply by LordOfBunnies** [reply]
The second one is more accurate to what I'm doing.  I only have the files open when I'm writing into them. I'm not as concerned about the performance issue as I am about it crashing.  I figure performance issues can be handled later when everything runs nicely. Is there a way to handle the metadata issues?  Why would that be causing it to crash?

The issue (using Backtrace) seems to be coming from hdf5, as the Backtrace routes from CGNS files into hdf5 files.

**Reply by brtnfld** [reply]
So you are packing all the output into a single CGNS file, and you don't want to leave the file open in case your code crashes. Hmm, maybe we should reintroduce the flush APIs. Unfortunately, metadata issues can not really be circumvented.

Is it still a problem if you don't close the file until all the data is written?

FYI, this is a similar report for NetCDF4, and we have yet to track down the issue, so hard to say what the issue is. You can also try with mpich. 

1,1,                        nc_enddef (dfile.c:1004)
1,1,                          NC4__enddef (hdf5file.c:507)
1,1,                            NC4_enddef (hdf5file.c:544)
1,1,                              nc4_enddef_netcdf4_file (hdf5file.c:775)
1,1,                                sync_netcdf4_file (hdf5file.c:212)
1,1,                                  nc4_rec_write_metadata (nc4hdf.c:1950)
1,1,                                    write_var (nc4hdf.c:1669)
1,1,                                      var_create_dataset (nc4hdf.c:1011)
1,1,                                        H5Dcreate2 (H5D.c:144)
1,1,                                          H5D__create_named (H5Dint.c:328)
1,1,                                            H5L_link_object (H5L.c:1556)
1,1,                                              H5L__create_real (H5L.c:1797)
1,1,                                                H5G_traverse (H5Gtraverse.c:850)
1,1,                                                  H5G__traverse_real (H5Gtraverse.c:626)
1,1,                                                    H5L__link_cb (H5L.c:1640)
1,1,                                                      H5G_obj_insert (H5Gobj.c:595)
1,1,                                                        H5G__dense_insert (H5Gdense.c:426)
1,1,                                                          H5HF_insert (H5HF.c:392)
1,1,                                                            H5HF__man_insert (H5HFman.c:137)
1,1,                                                              H5HF__man_dblock_new (H5HFdblock.c:397)
1,1,                                                                H5HF__hdr_update_iter (H5HFhdr.c:1054)
1,1,                                                                  H5HF__man_iblock_root_double (H5HFiblock.c:580)
1,1,                                                                    H5AC_move_entry (H5AC.c:1140)
1,1,                                                                      H5AC__log_moved_entry (H5ACmpio.c:1102)
1,1,                                                                        __assert_fail
1,1,                                                                          __assert_fail_base
1,1,                                                                            abort

**Reply by LordOfBunnies** [reply]
I think you've answered my question enough to get me going.  I'll just space out the outputs until I'm more confident in my code, then leave the output files open until it's time to finalize everything.

I'm still unsure of the exact cause, but you've given me enough to work around the issue.  Thank you!
